Jint/Native/Proxy/JsProxy.cs

```diff
@@ -134,25 +135,55 @@ public override JsValue Get(JsValue property, JsValue receiver)
             var targetDesc = target.GetOwnProperty(property);
             if (targetDesc != PropertyDescriptor.Undefined)
             {
-                if (targetDesc.IsDataDescriptor() && !targetDesc.Configurable && !targetDesc.Writable && !ReferenceEquals(result, targetDesc._value))
+                if (targetDesc.IsDataDescriptor())
                 {
-                   ExceptionHelper.ThrowTypeError(_engine.Realm);
+                    var targetValue = targetDesc.Value;
+                    if (!targetDesc.Configurable && !targetDesc.Writable && !IsSameValue(result, targetValue))
+                    {
+                        ExceptionHelper.ThrowTypeError(_engine.Realm, $"'get' on proxy: property '{property}' is a read-only and non-configurable data property on the proxy target but the proxy did not return its actual value (expected '{targetValue}' but got '{result}')");
+                    }
                 }
-                if (targetDesc.IsAccessorDescriptor() && !targetDesc.Configurable && (targetDesc.Get ?? Undefined).IsUndefined() && !result.IsUndefined())
+
+                if (targetDesc.IsAccessorDescriptor())
                 {
-                   ExceptionHelper.ThrowTypeError(_engine.Realm, $"'get' on proxy: property '{property}' is a non-configurable accessor property on the proxy target and does not have a getter function, but the trap did not return 'undefined' (got '{result}')");
+                    if (!targetDesc.Configurable && (targetDesc.Get ?? Undefined).IsUndefined() && !result.IsUndefined())
+                    {
+                        ExceptionHelper.ThrowTypeError(_engine.Realm, $"'get' on proxy: property '{property}' is a non-configurable accessor property on the proxy target and does not have a getter function, but the trap did not return 'undefined' (got '{result}')");
+                    }
                 }
             }
 
             return result;
         }
 
+        private static bool IsSameValue(JsValue expected, JsValue actual)
```

Comment:
there's already `JsValue.IsSameValue` method that probably should be used, it doesn't handle `ObjectWrapper` though